packages/suite/src/components/suite/NumberInput.tsx

```diff
@@ -0,0 +1,323 @@
+            e.clipboardData.setData('text/plain', cleanValueString(displayValue, locale));
+
+            e.preventDefault();
+        },
+        [displayValue, locale],
+    );
+
+    const handleCut = useCallback(
+        (e: React.ClipboardEvent<HTMLInputElement>) => {
+            handleCopy(e);
+            handleChange('');
+        },
+        [handleCopy, handleChange],
+    );
+
+    // only allow inputs and separators
+    const handleOnBeforeInput = useCallback(
+        (e: React.FormEvent<HTMLInputElement> & { data: string }) => {
+            if (/[\d.,]/g.test(e.data)) {
+                return;
+            }
+
+            e.preventDefault();
+        },
+        [],
+    );
+
+    // checks for separators at pos + cursorCharacterOffset and moves the cursor to pos + cursorPositionOffset
+    const handleCursorShift = useCallback(
+        (cursorCharacterOffset: number, cursorPositionOffset: number) => {
+            if (!inputRef.current) {
+                return;
+            }
+
+            const { selectionStart: cursorPosition, value } = inputRef.current;
+            if (cursorPosition === null) {
+                return;
+            }
+
+            const { groupSeparatorCharCode } = getLocaleSeparatorCharCodes(locale);
+            const characterBerofeCursor = value[cursorPosition + cursorCharacterOffset];
+
+            if (characterBerofeCursor?.charCodeAt(0) === groupSeparatorCharCode) {
+                const newPosition = cursorPosition + cursorPositionOffset;
+                inputRef.current.setSelectionRange(newPosition, newPosition);
+            }
+        },
+        [inputRef, locale],
+    );
+
+    // do not allow putting the cursor after group separators
+    const handleClick = useCallback(() => {
+        handleCursorShift(-1, -1);
+    }, [handleCursorShift]);
+
+    // jump over group separators when navigatind the input
+    const handleKeyNav = useCallback(
+        (e: React.KeyboardEvent<HTMLInputElement>) => {
+            const pressedKey = e.key;
+
+            if (pressedKey === 'ArrowRight') handleCursorShift(0, 1);
+            if (pressedKey === 'ArrowLeft') handleCursorShift(-2, -1);
+        },
+        [handleCursorShift],
+    );
+
+    const handleKeyDown = useCallback(
+        (e: React.KeyboardEvent<HTMLInputElement>) => {
+            const pressedKey = e.key;
+
+            if (['ArrowLeft', 'ArrowRight'].includes(pressedKey)) {
+                handleKeyNav(e);
+                return;
+            }
+
+            // undo doesn't work properly so better disallow it altogether
+            if ((e.ctrlKey || e.metaKey) && e.key === 'z') {
+                e.preventDefault();
+            }
+        },
+        [handleKeyNav],
+    );
+
+    return (
+        <Input
+            {...props}
+            {...controlProps}
+            innerRef={inputRef}
+            value={displayValue}
+            inputMode="decimal"
+            onClick={handleClick}
```

Comment:
If you select everything and then click behind the separator, the cursor will stay there instead of moving before the separator. This is because the `onclick` handler is not aware of the current selection. You can fix that be changing the event to `onSelect`.

Reply:
a6ef76441